src/game.cpp

```diff
@@ -6494,60 +6497,87 @@ void game::explosion( const tripoint &p, float power, float factor,
-        // TODO: Z-level shrapnel, but not before z-level ranged attacks
-        tripoint sp{ static_cast<int> (rng( p.x - radius, p.x + radius )),
-                     static_cast<int> (rng( p.y - radius, p.y + radius )),
-                     p.z };
 
-        proj.impact = damage_instance::physical( power, power, 0, 0 );
+    for( auto i = 0; i != count; ++i ) {
+        // shrapnel expands randomly in all directions
+        tripoint dst = m.random_perimeter( src, range );
 
-        Creature *critter_in_center = critter_at( p ); // Very unfortunate critter
-        if( critter_in_center != nullptr ) {
-            dealt_projectile_attack dda; // Cool variable name
-            dda.proj = proj;
-            // For first shrapnel piece:
-            // 50% chance for 50%-100% base (power to 2 * power)
-            // 50% chance for 0-25% base
-            // Each one after that gets a progressively lower chance of hitting
-            dda.missed_by = rng_float( 0.4, 1.0 ) + (i * 1.0 / count);
-            critter_in_center->deal_projectile_attack( nullptr, dda );
-        }
+        int kinetic = power;
+        bresenham( src, dst, 0, 0, [this,&distrib,&i,&src,&kinetic,&mass,&proj]( const tripoint& e ) {
 
-        if( sp != p ) {
-            // This needs to be high enough to prevent game from thinking that
-            //  the fake npc is scoring headshots.
-            fake_npc.projectile_attack( proj, sp, 3600 );
-        }
+            distrib[ e ] += 0; // add this tile to the distribution
+
+            auto critter = critter_at( e );
+            if( critter && !critter->is_dead_state() ) {
+                // special case critter at epicenter to have equivalent chance to adjacent tile
+                if( src == e && !one_in( 8 ) ) {
+                    return true;
+                }
+
+                dealt_projectile_attack frag;
+                frag.proj = proj;
+                frag.missed_by = 0;
+                frag.proj.impact = damage_instance::physical( 0, kinetic, 0, std::min( kinetic, mass ) );
+
+                distrib[ e ] += kinetic; // increase received damage for tile in distribution
+
+                add_msg( m_debug, "fragment %i hit %s with kinetic=%i", i, critter->get_name().c_str(), kinetic );
+                critter->deal_projectile_attack( nullptr, frag );
+                return false;
+            }
+
+            if( m.impassable( e ) ) {
+                // massive shrapnel can smash a path through obstacles
+                int resistance = m.bash_resistance( e );
+                if( m.bash( e, std::min( kinetic, mass ), true ).success ) {
+                    distrib[ e ] += resistance; // obstacle absorbed only some of the force
+                    kinetic -= resistance;
+                } else {
+                    distrib[ e ] += kinetic; // obstacle absorbed all of the force
+                    return false;
+                }
+            }
+
+            kinetic -= m.move_cost( e );
```

Comment:
At the moment we don't have much real cover. Pretty much single tiles at most (things like windows). Except cars.

Did you test cars? Something tells me the new shrapnel will fly through cars as if they weren't there.